File: Array/1962.Remove_Stones_to_Minimize_the_Total/Python/1962.Remove_Stones_to_Minimize_the_Total.py
class Solution(object):
    def minStoneSum(self, piles, k):
        """
        :type piles: List[int]
        :type k: int
        :rtype: int
        """
        #============================================#
        # Heap/Priority-queue based traversal method #
        #============================================#

        ############
        #Initialize
        ##### Length of piles array #####
        len_piles = len(piles)

        ##### Updated/Heapify piles array #####
        for piles_idx in range(len_piles):
            piles[piles_idx] *= (-1)

        heapify(piles)

        
        ##########################################
        #Heap/Priority-queue based loop traversal
        for k_idx in range(k):
            curr_min_val = heappop(piles) #Current minimun value

            ##### Check if the current indexed-value is divisable by 2 or not #####
            if (not (curr_min_val % 2)):
                heappush(piles, (curr_min_val - (curr_min_val // 2))) #Keep updating/recording
            else:
                heappush(piles, (curr_min_val - ((curr_min_val // 2) + 1))) #Keep updating/recording

        return (sum(piles) * (-1))

File: Array/1962.Remove_Stones_to_Minimize_the_Total/C++/1962.Remove_Stones_to_Minimize_the_Total.cpp
class Solution {
public:
    int minStoneSum(vector<int> & piles, int k) 
    {
        //==========================================//
        // Input type:                              //
        //   - vector<int> ref. (piles vector)      //
        //   - int (k-th operations)                //
        // Return type:                             //
        //   - int (result minimun summary counter) //
        //==========================================//

        /*Initialize*/
        ///// Record heap, priority-queue /////
        priority_queue<int> record_pq(begin(piles), end(piles));

        ///// Record maximun value /////
        int record_max_val;

        ///// Result minimun summary counter /////
        int res_min_cnter = accumulate(piles.begin(), piles.end(), 0);


        /*Heap, Priority-queue based loop traversal*/
        for (int k_idx = 0; (k_idx < k); (k_idx++)) //K-th operations
        {
            record_max_val = record_pq.top(); //Record maximun value
            record_pq.pop(); //Keep updating, popping

            record_pq.push((record_max_val - (record_max_val / 2))); //Keep updating, recording

            res_min_cnter -= (record_max_val / 2); //Keep updating, reducing

        } //K-th operations

        return res_min_cnter;
    }
};
